<!DOCTYPE html>
<html>
<head>
    <title>KUBE Ajax: Client Object and API</title>
    <!-- Include KUBE.js -->
    <script src="../../../src/KUBE.js"></script>

    <!-- /Library/Ajax/Client Object API and Usage -->
    <script type="text/javascript">
        /*
         *  The Client object is used to communicate with a target server (usually remote though it can be local)
         *
         *  Current API reference for /Library/Ajax/Client
         *      -SetTarget
         *      -SetTimeout
         *      -SendRequest
         *      -CreateRequest
         */

        //Load our Ajax/Client in
        KUBE.Uses('/Library/Ajax/Client',clientExample);
        function clientExample(_ClientDef){

            //Create a new Client
            var Client = _ClientDef();

            //SetTarget: This can be a URL (string) or a function callback (which we will not look into in this example, see Client: Implementing a Local Client example)
            //  -Also we will look at a more indepth example of this in Implementing a Remote Client example
            Client.SetTarget('http://google.com');

            //SetTimeout: This takes in milliseconds and sets an automatic Timeout on every request. In the event of a Timeout, the Promise ends up Rejected
            Client.SetTimeout(10000);

            //CreateRequest: This is just a wrapper to /Library/Ajax/Request (see Request Example)
            var NewRequest = Client.CreateRequest();
            NewRequest.AddHeader('test','test');
            NewRequest.AddData('key','val');

            //SendRequest: Only accepts a Request Object (/Library/Ajax/Request) and returns a KUBE.Promise() object (see Core: KUBE.Promise() example)
            var WaitForResponse = Client.SendRequest(NewRequest);

            //The internal execute can only resolve once a valid Response object has been formed and returned. It is the third argument passed into the promise
            WaitForResponse.Then(function(_resolve,_reject,_Response){
                console.log('Response Received');
            });

            //Or it can be rejected in which case an _err object is returned with _err.msg and _err.data)
            WaitForResponse.Catch(function(_resolve,_reject,_err){
                //In this case we will have run into Cross Origin Request, which results in an Network failure (timeout) which will Reject the promise
                console.log(_err.msg);
            });
        }

    </script>

</head>
<body>
Example outputs to console:
</body>
</html>